### PR TITLE
Add logger

### LINKED ISSUE #33: Add logging capabilities
**Is your feature request or enhancement related to a problem? What specifically could be better? Please describe.**
We shouldn't use std::cout or std::cerr to log/print messages to the console. This doesn't give us control over what messages are seen (eg. seeing all messages, just debug messages, etc.). Similarly, we should not directly use `ROS_DEBUG` (or similar `ROS_` macros) so we can stay decoupled from the ROS framework.

We should still have a way of logging messages to the console, and controlling what "level" of messages can be seen.

**Describe the solution you'd like**
A logging solution that lets us log messages to the console (and maybe a file) by severity level (INFO, DEBUG, WARN, ERR etc.).

**Describe alternatives you've considered**
Some options are:
* https://github.com/google/glog
* https://github.com/gabime/spdlog
* https://github.com/KjellKod/g3log (a faster version of glog),
* There are definitely other solutions out there. Bonus if the tool/library is easy to install via `apt-get` or `snap` on linux so it can easily be incorporated into our setup scripts

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description
*Give a high-level description of the changes in this PR*
Added [g3log](https://github.com/KjellKod/g3log) to use as the logger for our AI in place of the rosconsole API, so that we can log in a ROS-agnostic way.

Logs are output both to `stdout` so they show up in the console, as well as sent to `/rosout` via a custom log sink.

*Note*: The `std::cout` statements in the `main` functions for unit tests were not replaced, since that would require linking and initializing the logger for every test executable, and we want to keep the unit tests as simple and minimal as possible.

### Testing Done
*Outline any testing that was done for these changes. This could be unit tests, integration tests, etc.*
* Manually verified that log messages show up in the console
* Manually verified that log messages how up on the `/rosout` topic using `rqt_console`

### Resolved Issues
*Link any issues that this PR resolved. Eg `resolves #1, #2, and #5` (note that they MUST be specified like this so Github can automatically close them then this PR merges)*

resolves #33 

### Review Checklist
*(Please check every item to indicate your code complies with it (by changing `[ ]`->`[x]`). This will hopefully save both you and the reviewer(s) a lot of time!)*
***It is the reviewers responsibility to also make sure every item here has been covered***
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

***Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!***

### CODE REVIEW

File: environment_setup/setup_software.sh
Comment (EvanMorcom) [inline environment_setup/setup_software.sh:236]: Should the URL be some kind of constant string at the top of the file?

File: src/thunderbots/software/util/logger/custom_g3log_sinks.h
Comment (EvanMorcom) [inline src/thunderbots/software/util/logger/custom_g3log_sinks.h:1]: nit: no header comment

**Review by EvanMorcom — COMMENTED** [review]
Other than a couple nit items, LGTM 👍

**Comment by EvanMorcom** [comment]
Just thought, do we not need unit tests for this to make sure it works in our implementation?

**Comment by MathewMacDougall** [comment]
Hard to unit test logging like this, since once we send to `stdou`t we just have to trust that the OS handles it properly. Likewise there's not point in re-unit testing the library. If we want to be extra, we could have an integration test that makes sure the log messages get published to the `/rosout` topic, but I don't think we can unit test any other functionality.

@garethellis0 for his opinion.

File: environment_setup/setup_software.sh
Comment (MathewMacDougall) [reply environment_setup/setup_software.sh:236]: Don't think so. All our other scripts keep the urls and variables with the sections that use them. Leaving this as it is.

File: src/thunderbots/software/util/logger/custom_g3log_sinks.h
Comment (MathewMacDougall) [reply src/thunderbots/software/util/logger/custom_g3log_sinks.h:1]: fixed

File: src/thunderbots/software/ai/main.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/main.cpp]: Do we have to go through the process of setting up the logger in basically the same way for `ai`, `backend_input`, `backend_output`? Would be good to just have a function in `util` or something that did this for us if so

File: src/thunderbots/software/ai/main.cpp
Comment (garethellis0) [inline src/thunderbots/software/ai/main.cpp:113]: Nice, I like this style of logging!

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Overall looks really good. I like the selection of library. Easy to use and extensible. :+1: 

I seem to remember that there are a few places in the code that we should be using that. Is the intent to do that as part of #33? Or another ticket? If it is part of #33, is the plan just to open another PR with all the small replacements? (I very much support this if so)

**Comment by garethellis0** [comment]
I'd like to see a single rostest for this if possible, mostly as a sanity check. If it's at all a hassle just ping me and I can probably knock one up quick.

**Comment by garethellis0** [comment]
Changes look good, LGTM, pending a quick ROSTest as a sanity check

**Review by jessicahe21 — COMMENTED** [review]
Would differentiating Ubuntu distros (`xenial` vs `bionic` as mentioned in #109 ) be needed for setting up misc. utilities in `setup_software.sh`?

File: src/thunderbots/software/CMakeLists.txt
Comment (henryymliu) [inline src/thunderbots/software/CMakeLists.txt:105]: Small nit, but it may be a good idea to add this to something like a ${COMMON_LIBS} variable; that way it should save some work if we add more libraries that are required by all nodes in the future.

File: src/thunderbots/software/util/logger/init.h
Comment (henryymliu) [inline src/thunderbots/software/util/logger/init.h]: Another small nit, but is it worth mentioning that this self-destroys after scope exit? I was quite confused on why there wasn't a deinit function until reading the docs.

File: src/thunderbots/software/CMakeLists.txt
Comment (henryymliu) [inline src/thunderbots/software/CMakeLists.txt:48]: Would be nice if we fail early if the library wasn't found.

File: src/thunderbots/software/CMakeLists.txt
Comment (henryymliu) [inline src/thunderbots/software/CMakeLists.txt:47]: These messages seem like they're there for debugging; consider removing?

**Review by henryymliu — CHANGES_REQUESTED** [review]
Looks pretty good over, some small requests.

File: src/thunderbots/software/CMakeLists.txt
Comment (garethellis0) [reply src/thunderbots/software/CMakeLists.txt:48]: I believe this will fail on this line if the library is not found.

**Comment by MathewMacDougall** [comment]
@garethellis0 I did a quick search of the codebase for `cout` and `cerr`, and only found one instance to change in the `ai` node. The rest are in the `main` functions of unit tests, which I decided not to change for reasons I noted in the description of this PR. So I think everything we need to replace is replaced, at least that exists in `master` currently.

**Comment by MathewMacDougall** [comment]
@jessicahe21 I don't think we need to differentiate distros for what's currently in the "misc utilities". Cloning and building `g3log` from its repository will work across all distros, and since no one has issues with the misc utilities when everyone was getting set up at the start of the year, I believe those packages have consistent names across multiple distros and therefore don't need to be differentiated. Nice job being thorough though, that's a good situation to be aware of :+1: 

Your understanding of the logger is correct. Using a logger rather than `std::cout` gives us more control over the messages, and we use a 3rd party library rather than the ROS macros (like `ROS_INFO`) so that it is easy to move frameworks in the future if we no longer use ROS for whatever reason.

Yes, the log messages _can_ be used for debugging. I'm not sure exactly how we'll use the `DEBUG` log level in our main code, but we do have the option available. It's a good option even when debugging locally, since when the code is build in `debug` mode the log statements will be compiled-in and run, but when the code is built in `release` the debug log statements aren't compiled in at all, meaning they won't show up without slowing down the program. This would save you time adding/removing debug statements as you were testing something.

We haven't used a "real" logger like this previously so we still need to decide how things like the `DEBUG` log level will be used. That's something good to think about and ask others what they think.

File: src/thunderbots/software/CMakeLists.txt
Comment (henryymliu) [reply src/thunderbots/software/CMakeLists.txt:48]: I thought it would just set `${G3_LIBRARY_PATH}` to be empty? https://cmake.org/cmake/help/v3.4/command/find_library.html

Comment (garethellis0) [reply src/thunderbots/software/CMakeLists.txt:48]: Oh sorry, you're completely correct (just tested this out on my local machine to confirm). I was thinking of `find_package`.

**Review by matthewberends — DISMISSED** [review]

**Comment by garethellis0** [comment]
@MathewMacDougall - any progress on that ROSTest?

File: src/thunderbots/software/util/logger/init.h
Comment (MathewMacDougall) [reply src/thunderbots/software/util/logger/init.h]: This is now documented with the LoggerSingleton class

File: src/thunderbots/software/CMakeLists.txt
Comment (MathewMacDougall) [reply src/thunderbots/software/CMakeLists.txt:105]: I'm not going to worry about this until we add more libs. In the future if we ended up with a small number of common libs I might vouch for keeping them separate anyway for the sake of explicitness, but that's a discussion for later.

File: src/thunderbots/software/CMakeLists.txt
Comment (MathewMacDougall) [reply src/thunderbots/software/CMakeLists.txt:47]: I'm thinking of leaving them in because they are similar to the protobuf messages like `Found PROTOBUF Compiler: /usr/bin/protoc` that I find useful as status reports as CMake runs. Do you have a strong preference?

**Comment by MathewMacDougall** [comment]
The ROSTests are finally working!!! :relieved:

File: src/thunderbots/software/CMakeLists.txt
Comment (MathewMacDougall) [reply src/thunderbots/software/CMakeLists.txt:48]: fixed

File: src/thunderbots/software/test/test_util/rostest_util.h
Comment (garethellis0) [inline src/thunderbots/software/test/test_util/rostest_util.h]: Could we rename this to `RosTestUtil` to:
a) better match the file name
b) be less likely to conflict with namespaces from other projects

File: src/thunderbots/software/test/util/logger_test_node.cpp
Comment (garethellis0) [inline src/thunderbots/software/test/util/logger_test_node.cpp]: Hmm, after looking at the repo we're not _super_ consistent with naming for test files, but could we rename this to `logger_test` or rename the class in this file to `LoggerTestNode` and rename the other RosTest files and classes accordingly? I'd be happy with either option

**Comment by garethellis0** [comment]
Nice job getting those RosTest utils done. They should be _really_ useful in the future!

**Comment by garethellis0** [comment]
@henrymliu - looks like this needs a re-review from ya

File: src/thunderbots/software/test/util/logger_test_node.cpp
Comment (MathewMacDougall) [reply src/thunderbots/software/test/util/logger_test_node.cpp]: Renamed to `logger_test`

File: src/thunderbots/software/test/test_util/rostest_util.h
Comment (MathewMacDougall) [reply src/thunderbots/software/test/test_util/rostest_util.h]: fixed

File: src/thunderbots/software/test/test_util/rostest_util.h
Comment (MathewMacDougall) [reply src/thunderbots/software/test/test_util/rostest_util.h]: fixed

**Review by garethellis0 — APPROVED** [review]

**Comment by garethellis0** [comment]
LGTM :+1: :tada:

File: src/thunderbots/software/CMakeLists.txt
Comment (henryymliu) [reply src/thunderbots/software/CMakeLists.txt:105]: Ok 👍

**Review by henryymliu — APPROVED** [review]
LGTM

**Pull request merged**
